src/ustar_header_parser.e

```diff
@@ -19,169 +20,212 @@ inherit
 			end
 
-			-- parse linkname
-			current_field := parse_string (block, pos + {TAR_CONST}.tar_header_linkname_offset, {TAR_CONST}.tar_header_linkname_length)
-			if (attached current_header as header and not current_field.is_empty) then
-				header.set_linkname (create {PATH}.make_from_string (current_field))
+				-- parse uname
+			if not err then
+				l_field := next_block_string (block, pos + {TAR_CONST}.tar_header_uname_offset, {TAR_CONST}.tar_header_uname_length)
+				if not l_field.is_whitespace then
+					l_header.set_user_name (l_field)
+				else
+--					err := True
+				end
 			end
 
-			-- parse and check magic
-			current_field := parse_string (block, pos + {TAR_CONST}.tar_header_magic_offset, {TAR_CONST}.tar_header_magic_length)
-			if (not (current_field ~ {TAR_CONST}.ustar_magic)) then
-				current_header := Void
+				-- parse gname
+			if not err then
+				l_field := next_block_string (block, pos + {TAR_CONST}.tar_header_gname_offset, {TAR_CONST}.tar_header_gname_length)
+				if not l_field.is_whitespace then
+					l_header.set_group_name (l_field)
+				else
+--					err := True
+				end
 			end
 
-
-			-- parse and check version
-			current_field := parse_string (block, pos + {TAR_CONST}.tar_header_version_offset, {TAR_CONST}.tar_header_version_length)
-			if (not (current_field ~ {TAR_CONST}.ustar_version)) then
-				current_header := Void
+				-- parse devmajor
+			if not err then
+				l_field := next_block_octal_string (block, pos + {TAR_CONST}.tar_header_devmajor_offset, {TAR_CONST}.tar_header_devmajor_length)
+				if l_field /= Void then
+					l_header.set_device_major (octal_string_to_natural_32 (l_field))
+				else
+					err := True
+				end
 			end
 
-			-- parse uname
-			current_field := parse_string (block, pos + {TAR_CONST}.tar_header_uname_offset, {TAR_CONST}.tar_header_uname_length)
-			if (attached current_header as header and not current_field.is_empty) then
-				header.set_user_name (current_field)
+				-- parse devminor
+			if not err then
+				l_field := next_block_octal_string (block, pos + {TAR_CONST}.tar_header_devminor_offset, {TAR_CONST}.tar_header_devminor_length)
+				if l_field /= Void then
+					l_header.set_device_minor (octal_string_to_natural_32 (l_field))
+				else
+					err := True
+				end
 			end
 
-			-- parse gname
-			current_field := parse_string (block, pos + {TAR_CONST}.tar_header_gname_offset, {TAR_CONST}.tar_header_gname_length)
-			if (attached current_header as header and not current_field.is_empty) then
-				header.set_group_name (current_field)
-			end
+				-- TODO: parse prefix
+				-- FIXME: Implement filename splitting
 
-			-- parse devmajor
-			current_field := parse_string (block, pos + {TAR_CONST}.tar_header_devmajor_offset, {TAR_CONST}.tar_header_devmajor_length)
-			if (not is_octal_natural_32_string (current_field)) then
-				current_header := Void
+			if not err then
+				last_parsed_header := l_header
 			end
+			parsing_finished := True
```

Comment:
I intended to set `last_parsed_header` to `Void` on parse failure instead of an attribute (that is the reason I made it detachable)